src/bindings/swig/ruby/kdb.i

```diff
@@ -0,0 +1,691 @@
+
+/* ignore unused constructors */
+%ignore kdb::KeySet::KeySet (ckdb::KeySet * k);
+%ignore kdb::KeySet::KeySet (size_t alloc, ...);
+%ignore kdb::KeySet::KeySet (VaAlloc alloc, va_list ap);
+
+%ignore kdb::VaAlloc;
+
+
+/* ignore raw ckdb::KeySet methods */
+%ignore kdb::KeySet::getKeySet;
+%ignore kdb::KeySet::setKeySet;
+%ignore kdb::KeySet::release;
+
+/* ignore unused operators */
+%ignore kdb::KeySet::operator=;
+/* KeySet == operator see below */
+%ignore kdb::operator== (const KeySet &, const KeySet &);
+%ignore kdb::operator!= (const KeySet &, const KeySet &);
+
+
+/*
+ * Constructors
+ */
+
+/* special mapping for KeySet::KeySet(Key, ...) constructor
+ * to enable passing a single Key, or an Array of Keys.
+ * This allows KeySet creation in a more Ruby way */
+/* first check if we've got a Key or a Ruby-array */
+%typemap(in) (kdb::Key, ...) {
+  $2 = NULL;
+  if (!RB_TYPE_P($input, T_ARRAY)){
+    if (SWIG_ConvertPtr($input, (void**)&$2, SWIGTYPE_p_kdb__Key, 0) == -1) {
+      rb_raise(rb_eArgError, "Argument has to be of Type Kdb::Key or Array");
+      SWIG_fail;
+    }
+  }
+}
+/* define a custom KeySet creation to be able to append the given Key 
+ * arguments to the newly created KeySet */
+%feature("except") kdb::KeySet::KeySet (Key, ...) {
+  /* original action
+  $action
+  */
+
+  if (arg2 != NULL) {
+    /* we got a kdb::Key argument (see corresponding typemap) */
+    kdb::Key *k = (kdb::Key *)arg2;
+    result = (kdb::KeySet *)new kdb::KeySet();
+    result->append(*k);
+  } else {
+    /* Ruby-Array */
+    if (RARRAY_LEN(argv[0]) > KEYSET_SIZE) {
+      /* if we know that the Array is bigger than the default KeySet size
+         create a KeySet which is able to hold all elements without 
+         reallocation */
+      result = (kdb::KeySet *)new kdb::KeySet(RARRAY_LEN(argv[0]), KS_END);
+    } else {
+      result = (kdb::KeySet *)new kdb::KeySet();
+    }
+    /* append each array element, while checking if we really got a Key */
+    for (int i = 0; i < RARRAY_LEN(argv[0]); i++) {
+      VALUE e;
+      kdb::Key *ek = NULL;
+      e = rb_ary_entry(argv[0], i);
+      if (SWIG_ConvertPtr(e, (void**)&ek, SWIGTYPE_p_kdb__Key, 0) == -1) {
+        rb_raise(rb_eArgError, 
+            "Array element at index %d is not of Type Kdb::Key", i);
+        delete result;
+        SWIG_fail;
+      }
+      result->append(*ek);
+    }
+  }
+  DATA_PTR(self) = result;
+}
+
+
+
+/* 
+ * KeySet.each
+ * Hint: this implementation of 'each' only works wich references to keys
+ * so any modifications of the keys are persisted
+ */
+%extend kdb::KeySet {
+  void each() {
+    if (rb_block_given_p()) {
+      cursor_t cur_pos = $self->getCursor();
+
+      for ( $self->rewind(); $self->next(); ) {
+        VALUE cur;
+        Key * t = new Key($self->current());
+        cur = SWIG_NewPointerObj(t, SWIGTYPE_p_kdb__Key, 1);
+
+        rb_yield(cur);
+
+        /* TODO: do we have to free anything ? */
```

Comment:
In this case, this isn't a memleak, since the new Ruby obj. will be the owner of the newly created key.
Latest valgrind runs didn't show any memleaks.